components/ContainedInputViews/src/MDCContainedInputView.h

```diff
@@ -0,0 +1,319 @@
+@property(strong, nonatomic, nonnull) UIColor *floatingLabelColor;
+@property(strong, nonatomic, nonnull) UIColor *placeholderColor;
+@property(strong, nonatomic, nonnull) UIColor *clearButtonTintColor;
+@property(strong, nonatomic, nonnull) UIColor *errorColor;
+@end
+
+@protocol MDCContainedInputViewStyleDensityInforming;
+
+@protocol MDCContainedInputViewStyle <NSObject>
+/**
+ The style's densityInformer. Different implementations of MDCContainedInputView may need different
+ density informers on a per style basis.
+ */
+@property(strong, nonatomic, nonnull) id<MDCContainedInputViewStyleDensityInforming>
+    densityInformer;
+/**
+ This method provides a default object conforming to MDCContainedInputViewColorScheming.
+ */
+- (nonnull id<MDCContainedInputViewColorScheming>)defaultColorSchemeForState:
+    (MDCContainedInputViewState)state;
+/**
+  This method allows objects conforming to MDCContainedInputViewStyle to apply themselves to objects
+  conforming to MDCContainedInputView with a set of colors represented by an object conforming to
+  MDCContainedInputViewColorScheming.
+ */
+- (void)applyStyleToContainedInputView:(nonnull id<MDCContainedInputView>)inputView
+    withContainedInputViewColorScheming:(nonnull id<MDCContainedInputViewColorScheming>)colorScheme;
+/**
+  This method allows objects conforming to MDCContainedInputViewStyle to remove the styling
+  previously applied to objects conforming to MDCContainedInputView.
+ */
+- (void)removeStyleFrom:(nonnull id<MDCContainedInputView>)containedInputView;
+/**
+  The value returned by this method determines how big the font of the floating label should be when
+  it's floating.
+ */
+- (CGFloat)floatingFontSizeScaleFactor;
+@end
+
+/**
+ Objects conforming to this protocol provide information about the vertical positions of views. This
+ helps achieve the variations in floating label position across the filled and outlined styles as
+ well as the general density of the views.
+ */
+@protocol MDCContainedInputViewStyleDensityInforming <NSObject>
+/**
+ This is a value between 0 and 1 that determines the visual vertical density of the view.
+ */
+@property(nonatomic, assign) CGFloat verticalDensity;
+/**
+ This method returns the mininum Y coordinate of the floating label given its height.
+ */
+- (CGFloat)floatingLabelMinYWithFloatingLabelHeight:(CGFloat)floatingLabelHeight;
+/**
+ This method returns the top padding of the main content area (where the text is) when the floating
+ label is floating.
+ */
+- (CGFloat)contentAreaTopPaddingFloatingLabelWithFloatingLabelMaxY:(CGFloat)floatingLabelMaxY;
+/**
+ This method returns the vertical (top and bottom) padding of the main content area (where the text
+ is) when the floating label is in the normal (not floating) state.
+ */
+- (CGFloat)contentAreaVerticalPaddingNormalWithFloatingLabelMaxY:(CGFloat)floatingLabelMaxY;
+@end
+
+/**
+ A base implementation of MDCContainedInputViewStyle.
```

Comment:
Tip: What does the base implementation do? What is it for? Can you elaborate in the comments?

Reply:
The filled and outline styles are only achievable by calling the appropriate methods on a theming extension, but `containerStyle` is a nonnull property on MDCContainedInputView. So the base style is used until it's themed to look filled or outlined.